scripts/download_checkpoints.py

```
@@ -0,0 +1,139 @@
+    
+    with open(save_path, 'wb') as f:
+        with tqdm(total=total_size, unit='B', unit_scale=True, desc=save_path.name) as pbar:
+            for chunk in response.iter_content(chunk_size=chunk_size):
+                if chunk:
+                    f.write(chunk)
+                    pbar.update(len(chunk))
+
+def try_get_directory_listing(url):
+    common_files = [
+        "config.yaml",
+        "model.pt",
+        "optim.pt",
+        "train.pt",
+        "model.safetensors",
+        "optim.safetensors",
+    ]
+    
+    found_files = []
+    for pattern in common_files:
+        test_url = urljoin(url.rstrip('/') + '/', pattern)
+        try:
+            response = requests.head(test_url)
+            if response.status_code == 200:
+                found_files.append(pattern)
+        except requests.exceptions.RequestException:
+            continue
+            
+    return found_files
+
+def download_checkpoint(url, save_dir):
+    """Download all files from a checkpoint directory."""
+    r2_url = convert_to_r2_url(url)
+    public_url = convert_to_public_url(r2_url)
+    
+    base_path = Path(save_dir)
+    base_path.mkdir(parents=True, exist_ok=True)
+    
+    print(f"\nR2 URL: {r2_url}")
+    print(f"Public URL: {public_url}")
+    print(f"Saving to: {base_path}")
+    
+    print("Checking for available files...")
+    available_files = try_get_directory_listing(public_url)
+    
+    if not available_files:
+        print("No files found using common patterns. The directory might be empty or use different file patterns.")
+        return
+    
+    for file in available_files:
+        file_url = urljoin(public_url.rstrip('/') + '/', file)
+        file_path = base_path / file
+        
+        try:
+            print(f"\nDownloading: {file}")
+            download_file(file_url, file_path)
+        except requests.exceptions.RequestException as e:
+            print(f"Error downloading {file}: {e}")
+            continue
```

No, wait, don't just catch all exceptions. What about those that can be retried, like timeouts?
Also, there is a risk that this will just skip some files and you don't notice because it prints a tiny message among thousands.

---

For each step we have only 4 files(not a lot). You want me to add timeout exception and retry it again?

---

Added timeout exception with the help of GPT.